Blog-Post: Rust für Java Entwickler - Teil 1 (#123)
adessoAG/devblog: _posts/2019-12-20-Rust-für-Java-Entwickler-Teil-1.md

Im Fehlerfall könnte dann eine entsprechende Fehlermeldung zurückgegeben werden.
Auf eine ausgiebige und konsistente Fehlerbehandlung gehen wir hier nicht ein.

## Modifizieren einer Person

Wenn wir eine Person modifizieren wollen, dann wollen wir entweder seinen Namen, sein Alter oder beides ändern.
Auch hier wollen wir im Request wieder die Änderungen als JSON-Body darstellen.
Also müssen wir an dieser Stelle eine neue Struktur definieren, die unseren Anforderungen gerecht wird:

```rust
#[derive(Serialize, Deserialize)]
pub struct UpdatePerson {
    pub name: Option<String>,
    pub age: Option<u32>
}
```

Der einzige Unterschied zum NewPerson-Struct ist, dass wir den Namen bzw. das Alter nicht angeben *müssen*.
Unser Request Handler könnte dann so aussehen:

```rust
#[put("/persons/{id}")]
pub fn update(id: Path<u32>, person: Json<UpdatePerson>) -> String {
    // ...
}
```

## Löschen einer Person

Beim Löschen einer Person gibt es keine neuen Funktionen oder Techniken des Frameworks, die vorzustellen wären.
Deshalb kurz und knapp - hier ist unser Request Handler:

```rust
#[delete("/persons/{id}")]
pub fn delete(id: Path<u32>) -> String {
    // ...
}
```

> Wir haben nun alle Request Handler für unsere REST-API definiert.
> Die Kommunikation mit einer Datenbank haben wir der Übersichtlichkeit und Einfachheit halber weggelassen.
> Dazu müssten wir ein komplett neues Framework einbinden und vorstellen - das wäre ein Thema für einen weiteren Blog-Artikel.
> Doch abgesehen von der Persistenzschicht fehlen uns auch andere elementare Aspekte in unserer App, wie z.B. eine konsistente und durchgehende Fehlerbehandlung sowie Unit- und Integration-Tests.
> Die Behandlung dieser Themen würde allerdings den Rahmen des Artikels sprengen, weshalb ich an dieser Stelle nicht näher darauf eingehen werde.

Reviewer: ```suggestion
Wir haben nun alle Request Handler für unsere REST-API definiert.
```

# Fazit

> In diesem Artikel haben wir gesehen, wie wir eine REST-API mit dem Actix-Web Framework implementieren können.
> Dabei haben wir wichtige Funktionen des Frameworks kennengelernt: Request Handler, Routen, dynamische URL-Parameter und die Kommunikation mittels JSON.
> Wenn bereits Erfahrung in der Sprache Rust vorhanden ist, wird der Einstieg in Actix-Web nicht schwerfallen.
> Natürlich bietet das Framework noch einiges mehr an Funktionalität als in diesem Artikel beschrieben wurde.
> Die [offizielle Dokumentation](https://actix.rs/docs/) ist der beste Ort für diejenigen, die sich noch tiefer mit dem Framework beschäftigen wollen.
> Dort wird auch auf die komplette API-Dokumentation verwiesen.

Reviewer: ```suggestion
In diesem Artikel haben wir gelernt wie eine REST-API mit dem Actix-Web Framework implementiert werden kann.
Dabei haben wir bereits wichtige Funktionen des Frameworks verwendet: 
- Request Handler 
- Routen
- dynamische URL-Parameter 
- Kommunikation mittels JSON
Mit ein wenig Erfahrung in Rust ist der Einstieg also nicht sehr schwer.
Die [offizielle Dokumentation](https://actix.rs/docs/) ist die beste Anlaufstelle, um tiefer einzusteigen.

Natürlich bietet das Framework noch einiges mehr an Funktionalität als in diesem Artikel beschrieben wurde.
Weitere Aspekte wie die Kommunikation mit einer Datenbank, Fehlerbehandlung und Unit- bzw. Integration-Tests behalten wir uns für weitere Blog-Artikel vor.

Als Teaser finden interessierte Leser in diesem [Github-Repository](https://github.com/KilianKrause/rest-api-with-actix) eine erweiterte Version des Codes, die eine rudimentäre Kommunikation mit dem Dateisystem implementiert.

Ein Beispiel für eine vollständige App mit Actix-Web ist z.B. [hier](https://github.com/fairingrey/actix-realworld-example-app) zu finden.
Dort wird auch ein ORM für die Speicherung der Personen in einer Datenbank eingesetzt.
```

> Um sich auf das Wesentliche konzentrieren zu können, habe ich die Speicherung der Personen in diesem Artikel nicht implementiert.
> In diesem [Github-Repository](https://github.com/KilianKrause/rest-api-with-actix) ist eine erweiterte Version des Codes zu finden, wo zumindest eine rudimentäre Kommunikation mit dem Dateisystem implementiert ist.
> 
> Ein Beispiel für eine vollständige App mit Actix-Web ist z.B. [hier](https://github.com/fairingrey/actix-realworld-example-app) zu finden.
> Dort wird auch ein ORM für die Speicherung der Personen in einer Datenbank eingesetzt.

Reviewer: ```suggestion
```
